Add option to allow module_defaults to be merged with or overwrite an outer scope

s-hertel opened the pull request:
##### SUMMARY
Given a playbook setting variables for the same group at the play and task level, such as:
```
---
- hosts: localhost
  connection: local
  gather_facts: no
  module_defaults:
    group/aws:
      profile: shertel
  tasks:
  - name: block of tasks requiring a particular region
    module_defaults:
      group/aws:
        region: us-east-1
    block:
        ....
```
Within the block, the module_defaults that are available are {'group/aws': {'region': 'us-east-1'}} and profile from the outer scope is unavailable.

With `module_defaults_merge: True` set on the block:
```
---
- hosts: localhost
  connection: local
  gather_facts: no
  module_defaults:
    group/aws:
      profile: shertel
  tasks:
  - name: block of tasks requiring a particular region but still need profile
    module_defaults_merge: True
    module_defaults:
      group/aws:
        region: us-east-1
    block:
        ....
```
The module_defaults available within the block is`{'group/aws': {'region': 'us-east-1', 'profile': 'shertel'}}`.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
module_defaults

s-hertel committed:
Fix overwriting rather than updating module_defaults

s-hertel committed:
changelog

s-hertel commented:
Looking for a backwards compatible solution

s-hertel committed:
Add a new option module_defaults_merge with a corresponding MODULE_DEFAULTS_MERGE option in config

Add new keyword to the docs

Add tests

Add changelog

s-hertel committed:
pep8

s-hertel committed:
Remove special handling for module_defaults_merge
(fixed by 8c08d039896488f6694a54640a0bdd69baba5fa1)

agaffney commented on lib/ansible/executor/task_executor.py:
```diff
@@ -598,7 +598,14 @@ def _execute(self, variables=None):
         for default in self._task.module_defaults:
-            module_defaults.update(default)
+            if self._task.module_defaults_merge:
+                for group in default:
```
It would be helpful to have a comment here saying that this code block merges the "parent" and "child" defaults one level deeper in the structure for the purpose of merging arguments for individual modules/groups rather than overwriting at the module/group level.

s-hertel replied:
Thanks for reviewing, addressed.

agaffney reviewed (commented):
This looks good to me

s-hertel committed:
Add a comment for the new code block

Clarify documentation a little

s-hertel commented on lib/ansible/config/base.yml:
```diff
@@ -1493,6 +1493,19 @@ MAX_FILE_SIZE_FOR_DIFF:
+  env: [{name: ANSIBLE_MODULE_DEFAULTS_MERGE}]
+  ini:
+  - {key: module_defaults_merge, section: defaults}
+  yaml:
```
Going to remove this

s-hertel committed:
Remove yaml entry from config because bcoca has been removing them

s-hertel committed:
Instead, deprecate overwrite-by-default for module_defaults and add config-only option to enable merging

update changelog

append module_defaults warnings to task results for deduplication

s-hertel commented:
@agaffney After some discussion with @nitzmahone and @bcoca it was suggested I deprecate the default behavior of overwriting groups defined in different scopes rather than adding more play syntax. A module/group could still clear its defaults by using {}. Do you mind taking one more look at this?

s-hertel committed:
fix line length

mattclay commented:
I like making the merge behavior configurable and changing the default after a deprecation period. However, I don't like using an empty dictionary to clear the defaults for several reasons:

1. It's not possible to clear while also setting new values in the same scope.
1. It feels a bit unintuitive, especially when using a variable for the dictionary.
1. It doesn't seem consistent with the behavior of other keywords using dictionaries.

Perhaps we could use a more explicit sentinel value for clearing than `{}`. Maybe something like:

```yaml
module_defaults:
  ping:
    __clear: yes
    data: hello
```

The `__clear` key could be anything that should not conflict with an actual module parameter.

It may also help to point out that using anything other than a dictionary for `module_defaults` will currently traceback. This includes using no value for a group or module, such as:

```yaml
module_defaults:
  ping:
```

If we keep `{}` for clearing and we make `None` the same as `{}` then the above would also clear.

s-hertel committed:
Add tests for bad module_defaults values

s-hertel committed:
fix stat option, though assertion still blows up

s-hertel committed:
Check for a MutableMapping instead of dict

s-hertel committed:
Fix severity of error

s-hertel committed:
Simplify loops

s-hertel committed:
Use a '__clear' keyword instead of an empty dictionary

s-hertel committed:
Add some documentation for examples of the behavior added

s-hertel commented:
I do like the clearing syntax in 84bb20d and a29b331 better than {} and needing to set new values in an inner scope. @nitzmahone Do you have thoughts on it?

s-hertel committed:
Try to simplify the loop a little more

s-hertel committed:
pep8

mattclay commented on changelogs/fragments/51149-allow-merging-module-defaults-groups.yaml:
```diff
@@ -0,0 +1,6 @@
+    a new MODULE_DEFAULTS_MERGE config option. When the option is set to True, module_defaults defined in an
+    inner scope are merged with those in the outer scope. module_defaults for a module/group may be emptied
+    by providing an empty dictionary. In 2.12 the default of MODULE_DEFAULTS_MERGE will be changed from
+    False to True.
```
Update this to refer to the new `__clear__` key instead of an empty dictionary.

s-hertel replied:
Good catch. I cleaned up the changelog to use the `deprecated_features` section too.

s-hertel committed:
Fix the changelog

bcoca commented on changelogs/fragments/51149-allow-merging-module-defaults-groups.yaml:
```diff
@@ -0,0 +1,9 @@
+    the default behavior will be to merge module_defaults set in an inner scope with those set in the outer
+    scope. Set MODULE_DEFAULTS_MERGE to true to enable merging before 2.12.
+  - module_defaults - Using an empty dictionary to clear previous module defaults for a module/group is
+    deprecated and will be removed in 2.12. Instead, set the special '__clear__' key to true for the
```
instead of adding a new token, can't we use overwrite + explicitly null value?

having Sentinel class now lets us use null/None as a user provided 'nullifier'

s-hertel commented:
In an attempt to keep conversations about this in a central, accessible place I'm summarizing a conversation I had with @bcoca:

* Sentinel actually won't work since the module_defaults field attribute isn't defining the subdict contents
* Don't introduce new play syntax like `__clear__` for additive keywords
* Empty dict/null should be used for clearing
    * Unfortunately, this doesn't allow people to clear/set new module_defaults in the same scope and leads to lots of nested blocks
* Probably all additive keywords should be able to use this (e.g. tags, environment)
*  Don't deprecate overwrite-by-default

nitzmahone reviewed (commented):
Might want to put this on the agenda for Tuesday's IRC meeting, just to see if there's any other community feedback, but I personally much prefer this approach to new top-level syntax. It gets us on a path toward saner default behavior, while still providing a clean way to mask higher-scoped values when necessary.

Some minor bikeshedding might be worthwhile on a different name for `__clear__` since (to me anyway) that implies that you're mutating the outer scope more than preventing its values from being inherited. It's tough to come up with something that's not too wonky though (eg `mask` and `noinherit` would both fail that test IMO, `block` is already overloaded, so :man_shrugging:)

s-hertel commented:
@nitzmahone Thanks for taking a look. And already did so: https://github.com/ansible/community/issues/450#issuecomment-469292038 :smile:
